File: app/lib/src/transfer/transfer_chunk_policy.dart
import 'dart:math';

abstract final class TransferChunkPolicy {
  static const legacyBytes = 256 * 1024;
  static const preferredBytes = 1024 * 1024;

  static int negotiate(Object? advertisedBytes, {required int localMaximum}) {
    if (advertisedBytes is! int || advertisedBytes <= 0) {
      return min(localMaximum, legacyBytes);
    }
    return min(localMaximum, advertisedBytes);
  }
}

File: src/transfer/transfer-chunk-policy.ts
export const legacyTransferChunkBytes = 256 * 1024;
export const preferredTransferChunkBytes = 1024 * 1024;

export function negotiateTransferChunkBytes(
  advertisedBytes: number | undefined,
  localMaximum: number,
): number {
  if (
    advertisedBytes === undefined ||
    !Number.isSafeInteger(advertisedBytes) ||
    advertisedBytes <= 0
  ) {
    return Math.min(localMaximum, legacyTransferChunkBytes);
  }
  return Math.min(localMaximum, advertisedBytes);
}
